src/SkipStages.cpp

```diff
@@ -206,16 +204,52 @@ class PredicateFinder : public IRVisitor {
 
 class ProductionGuarder : public IRMutator {
 public:
-    ProductionGuarder(const string &b, Expr p): buffer(b), predicate(p) {}
+    ProductionGuarder(const string &b, Expr compute_p, Expr alloc_p):
+        buffer(b), compute_predicate(compute_p), alloc_predicate(alloc_p) {}
 private:
     string buffer;
-    Expr predicate;
+    Expr compute_predicate;
+    Expr alloc_predicate;
 
     using IRMutator::visit;
 
+    bool memoize_call_uses_buffer(const Call *op) {
+        internal_assert(op->call_type == Call::Extern);
+        internal_assert(starts_with(op->name, "halide_memoization"));
+        for (size_t i = 0; i < op->args.size(); i++) {
+            const Variable *var = op->args[i].as<Variable>();
+            if (var && 
+                starts_with(var->name, buffer + ".") && 
+                ends_with(var->name, ".buffer")) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void visit(const Call *op) {
+        // We need to guard call to halide_memoization_cache_lookup to only be executed
```

Comment:
The reason we still need to execute the lookup if the compute_predicate is false but the alloc_predicate is true is that in this case the consumer code is going to load from the buffer but not actually use the results, so we need some scratch memory to load from. For memoized Func, that memory is probably already in the cache, so we do the lookup to retrieve it rather than allocating new memory.

That reasoning is pretty subtle, so I'd expand the comment to include it.